connector/connection.go

```diff
@@ -0,0 +1,218 @@
+package connector
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"syscall"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/Allenxuxu/gev/connection"
+	"github.com/Allenxuxu/gev/eventloop"
+	"github.com/Allenxuxu/gev/log"
+	"github.com/Allenxuxu/gev/poller"
+	"github.com/RussellLuo/timingwheel"
+	"github.com/libp2p/go-reuseport"
+	"golang.org/x/sys/unix"
+)
+
+var ErrDialTimeout = errors.New("i/o timeout")
+
+type connectionSocketState uint8
+
+const (
+	connectingConnectionSocketState connectionSocketState = iota + 1
+	connectedConnectionSocketState
+	disconnectedConnectionSocketState
+)
+
+type Connection struct {
+	state connectionSocketState
+	loop  *eventloop.EventLoop
+	*connection.Connection
+	timeout time.Duration
+	result  chan error
+	fd      int
+
+	protocol connection.Protocol
+	tw       *timingwheel.TimingWheel
+	idleTime time.Duration
+	callBack connection.CallBack
+}
+
+func newConnection(
+	network, address string,
+	loop *eventloop.EventLoop,
+	timeout time.Duration,
+	protocol connection.Protocol,
+	tw *timingwheel.TimingWheel,
+	idleTime time.Duration,
+	callBack connection.CallBack) (*Connection, error) {
+
+	fd, err := unixOpenConnect(network, address)
+	if err != nil {
+		if fd > 0 {
+			_ = unix.Close(fd)
```

Comment:
I think `close(fd)` should be handle in function `unixOpenConnect `.
If `unixOpenConnect ` return error,  `fd` need be 0 （Close it inside the function `unixOpenConnect `）